helper/impls/helper-impl-111/src/main/java/com/linkedin/avroutil1/compatibility/avro111/codec/AliasAwareSpecificData.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2022 LinkedIn Corp.
+ * Licensed under the BSD 2-Clause License (the "License").
+ * See License in the project root for license information.
+ */
+
+
+package com.linkedin.avroutil1.compatibility.avro111.codec;
+
+import com.linkedin.avroutil1.compatibility.ClassLoaderUtil;
+import org.apache.avro.Schema;
+import org.apache.avro.specific.SpecificData;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class AliasAwareSpecificData extends SpecificData {
+
+    //same idea as the one in SpecificData
+    protected final static Map<String, Class<?>> CLASS_CACHE = new ConcurrentHashMap<>();
+    protected final static Class<?> NO_CLASS = (new Object() {}).getClass();
+
+    private final ClassLoader classLoader;//redefined here because older  1.7 versions dont actually have thisd
+
+    public AliasAwareSpecificData() {
+        this.classLoader = null;
+    }
+
+    public AliasAwareSpecificData(ClassLoader classLoader) {
+        super(classLoader);
+        this.classLoader = classLoader;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Object createEnum(String symbol, Schema schema) {
+        @SuppressWarnings("rawtypes")
+        Class c = getClass(schema);
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return Enum.valueOf(c, symbol);
+    }
+
+    @Override
+    public Object createFixed(Object old, Schema schema) {
+        Class<?> c = getClass(schema);
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return c.isInstance(old) ? old : newInstance(c, schema);
+    }
+
+    @Override
+    public Object newRecord(Object old, Schema schema) {
+        Class<?> c = getClass(schema); //this is what vanilla does
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return (c.isInstance(old) ? old : newInstance(c, schema));
+    }
+
+    protected Class<?> lookupByAlias(Schema namedSchema) {
+        Set<String> aliases = namedSchema.getAliases();
+        if (aliases != null) {
+            for (String alias : aliases) {
+                Class<?> byAlias = CLASS_CACHE.computeIfAbsent(alias, fqcn -> {
+                    try {
+                        return ClassLoaderUtil.forName(this.classLoader, alias);
+                    } catch (ClassNotFoundException e) {
+                        return null; //TODO- switch to NO_CLASS
```

Comment:
Why?

Reply:
a negative cache will optimize the failure scenario of no class found (repeatedly). its debatable whether its worth optimizing the failure case, but thats the intent here